### PR TITLE
Add and use new `Signals` struct

### PR DESCRIPTION
**Pull request by IanManske**
# Description
This PR introduces a new `Signals` struct to replace our adhoc passing around of `ctrlc: Option<Arc<AtomicBool>>`. Doing so has a few benefits:
- We can better enforce when/where resetting or triggering an interrupt is allowed.
- Consolidates `nu_utils::ctrl_c::was_pressed` and other ad-hoc  re-implementations into a single place: `Signals::check`.
- This allows us to add other types of signals later if we want. E.g., exiting or suspension.
- Similarly, we can more easily change the underlying implementation if we need to in the future.
- Places that used to have a `ctrlc` of `None` now use `Signals::empty()`, so we can double check these usages for correctness in the future.

### CODE REVIEW

File: crates/nu-protocol/src/engine/engine_state.rs [hunk 1]
```diff
@@ -177,6 +177,18 @@ impl EngineState {
         }
     }
 
+    pub fn interrupt(&self) -> &Interrupt {
```
Comment (rgwood) [inline crates/nu-protocol/src/engine/engine_state.rs:180]: The naming of this function is a little confusing IMO. When I see a _function_ named `interrupt()` called, I read that as a verb and assume that calling the function will interrupt something.

I wonder if something like `get_interrupt()` might be a little more clear (very open to other ideas too)

**Review by rgwood — APPROVED** [review]
Great change, thanks!

I think the naming might be a little confusing but it's not a blocker.

**Comment by IanManske** [comment]
Since we might add other kinds of signals in the future, I was thinking `Execution` might be a more general name instead of `Interrupt`. (Then, the other methods would be renamed appropriately.) Thoughts?

**Comment by rgwood** [comment]
Hmm. `Execution` feels a little vague IMO. If the struct is potentially going to be used for multiple kinds of signals, how about `Signals`?

**Comment by IanManske** [comment]
Ya, that sounds good :smile:

**Review by rgwood — APPROVED** [review]

**Pull request merged**